### ISSUE TITLE
[Bug]: Ascend-dispatched rotary path (Triton rope) corrupts Q/K norms for GLM-5.2 DSpark draft (bf16, NeoX, head_dim=192); forward_native restores accept_len 3.06 -> 5.55

### ISSUE THREAD

**Issue by Frozen7771** [vllm-project/vllm-ascend#12723]
### Your current environment

<details>
<summary>Environment (collected inside the serving container; the container is offline so collect_env.py could not be downloaded)</summary>

```text
Ascend A3 (16 dies), driver 25.5.1
Image: quay.io/ascend/vllm-ascend:nightly-releases-v0.23.0-a3

Python 3.12.13
torch 2.10.0+cpu   (standard Ascend arrangement: CPU torch build, NPU backend via torch_npu)
torch_npu 2.10.0.post2
vllm 0.23.0
vllm_ascend: source tree bundled in the image at /vllm-workspace/vllm-ascend (no __version__ attribute)
triton 3.2.0
CANN: as bundled in the image

vllm_ascend/ops/rotary_embedding.py: HAS_TRITON = True
-> rope_forward_oot routes into rope_forward_triton (vllm_ascend/ops/triton/rope.py),
   not torch_npu._npu_rotary_embedding, for this stack.
```

</details>

## Summary

While bringing up GLM-5.2 DSpark speculative decoding on Ascend A3 (16 dies, DP2/TP8/EP, eager), we found that the **platform-dispatched rotary embedding path returns non-norm-preserving Q/K** for the DSpark draft attention shape (bf16, NeoX-style, `head_dim = rotary_dim = 192`). Calling `forward_native()` on the **same rotary object, same worker, same inputs** produces correct results and restores online acceptance length from ~3.0 to ~5.4–5.5.

This is not a checkpoint / config / acceptance-logic problem (cf. #12262, which fixes different root causes — Markov logits input, anchor sampling, QuaRot weight rotation). It is a numerical error in the dispatched rotary execution path itself.

## Environment

- Ascend A3, 16 dies, driver 25.5.1
- Image: `quay.io/ascend/vllm-ascend:nightly-releases-v0.23.0-a3` (vllm-ascend bundled as a source tree in the image)
- Python 3.12.13, `vllm 0.23.0`, `torch 2.10.0+cpu` + `torch_npu 2.10.0.post2` (standard Ascend arrangement: CPU torch build with the NPU backend provided by torch_npu), `triton 3.2.0`, CANN as bundled in the image
- Target: GLM-5.2 (753B MoE), w8a8 (`is_rot_used=False`, so QuaRot rotation is not involved)
- Draft: self-trained GLM-5.2 DSpark speculator (bf16), k=7, offline mean accept_len ~5.6
- Serving code: PR #11066 (`Liuchenbing-2026:dspark` @ `fabc77a9`) applied onto the v0.23 image, plus minimal local shims needed to run that branch on this image (model registration, per-layer KV-cache spec dispatch, and one Triton-rope port described below). None of the shims changes RoPE math, weights, or head_dim.

The draft attention builds its rotary via vLLM `get_rope(head_dim=192, is_neox_style=True, ...)` and applies it as `q, k = self.rotary_emb(positions, q, k)`, i.e. through the vLLM `CustomOp` dispatch into the Ascend platform implementation.

## Which backend does the dispatch take?

On this stack, `vllm_ascend/ops/rotary_embedding.py` reports `HAS_TRITON = True`, and `rope_forward_oot` routes `if HAS_TRITON` into `rope_forward_triton` (`vllm_ascend/ops/triton/rope.py`), not into `torch_npu._npu_rotary_embedding`. So the corrupted path below is the **Triton rope path**.

One disclosure for reproducibility: the stock v0.23 Triton rope could not even compile this shape — with the large-head threshold at 256, `head_dim=192` selects `BLOCK_SIZE_HEAD=64`, which exceeds the BiSheng UB budget by ~2 KB. To run at all, we ported upstream main's `_triton_rope` / `rope_forward_triton` verbatim and only lowered the head tile for this shape (`BLOCK_SIZE_HEAD=16`). An isolated single-die test of the ported function on the exact shape (`nq=nk=64, hd=192`, cleared Triton cache) matches the PyTorch reference with **max error 0**. Despite that isolated exactness, the in-serving dispatched call is corrupted (evidence below) — i.e. the failure manifests only under real serving conditions (grid/parallelization configuration or the call contract), not in the kernel math tested in isolation.

Note that **upstream main has the same 256 threshold**, so `head_dim=192` draft shapes on main would select the same `BLOCK_SIZE_HEAD=64` tiling this hardware could not compile — maintainer confirmation of which rotary backend serves this shape on current main (post-#11431) would help scope the affected range.

## Symptom

With the same draft checkpoint:

- offline acceptance (teacher-forced, target-side scoring): mean accept_len ≈ 5.6
- online serving acceptance: **3.056** (8 serial requests) / **2.910** (32 requests, concurrency 8) — a global collapse, not a per-domain effect.

## Root-cause evidence chain

**1. First strong divergence is layer-0 query RoPE.** Comparing online layer-0 tensors against an offline replay of the same fixed prefix: input rel-L2 ≈ 0.0028, Q/K/V projections ≈ 0.0009, per-head RMSNorm outputs ≈ 0.003 — then **Q/K jump to rel-L2 0.68 / 0.70 immediately after RoPE**.

**2. RoPE must be norm-preserving; the dispatched path is not.** Online post/pre-RoPE norm ratios: **Q ≈ 0.737, K ≈ 0.732** (offline reference: 1.0001 / 0.9997). This magnitude cannot be bf16 rounding.

**3. cos/sin cache is exonerated.** Rebuilding the same `RotaryEmbedding(head_size=rotary_dim=192, is_neox=True)` cache: rows satisfy `cos² + sin² = 1` with max error 1.19e-7; CPU `forward_native` matches the wrapper bitwise.

**4. Same-worker A/B (key experiment).** Inside one live serving worker, on the same rotary object / cos-sin cache / positions / pre-RoPE Q/K, we saved three results side by side:
   - dispatched call vs. dispatched call on a `.contiguous()` clone: **bitwise identical** → input stride/layout is *not* the cause;
   - dispatched call vs. `forward_native()`: **rel-L2 = 0.678912 (Q) / 0.694659 (K)**, norm ratio **0.736589 / 0.732088**;
   - `forward_native()` preserves norms.

**5. Token-level causality, reproduced in two independent fresh server processes.** With a fixed greedy request, at a fixed draft position the proposal flips from a wrong token to exactly the target model's top-1 when the native-path result is selected — in both fresh runs.

**6. Production fix confirms global causality.** Changing only the draft query RoPE call to `self.rotary_emb.forward_native(positions, q, k)` (no other change, no diagnostic code):

| metric | before | after |
|---|---:|---:|
| accept_len, 8 serial | 3.055556 | **5.547692** |
| accept_len, 32 req / c8 | 2.910313 | **5.429936** |
| position-1 acceptance | 0.718 | 0.901 |
| position-7 acceptance | 0.024 | 0.394 |

All 40 requests HTTP 200, hard counters consistent (`draft_tokens = 7 × drafts`), post-run health green.

## What we have NOT concluded (scope honesty)

- We have **not** attributed the error to a specific kernel line / tiling / tail branch. The statement we can support is: the `CustomOp`-dispatched rotary runtime path (routed through the Triton rope on this stack) returns incorrect, non-norm-preserving Q/K for this shape under real serving, while `forward_native` on the same NPU is correct.
- We have **not** tested whether the post-#11431 refactored main path is affected. #12262 reports accept_len 4.95 on refactored main, which suggests either that path takes a different rotary branch for this shape, or the corruption is specific to this stack's dispatch.
- `forward_native()` runs on NPU as unfused tensor ops; end-to-end throughput comparison (no-spec vs MTP-5 vs DSpark) is in progress.

## Possibly related history

#7082 fixed a neox-style `rope_siso` result-storage bug causing accuracy degradation — precedent that the Ascend rotary custom paths have had numerical issues for NeoX-style shapes.

## Suggested fix / ask

Short term, a model-local fix (draft query RoPE → `forward_native`) restores correctness; we will submit it as a draft PR referencing this issue. Longer term this likely wants a shape-gated fix or fallback in the Ascend rotary backend itself (and possibly a revisit of the large-head tiling threshold for `head_dim=192`). We're happy to run further diagnostics on A3 — we have a same-worker A/B harness that can dump dispatched vs native results for any shape.

**Comment by yiz-liu** [comment]
@ZT-AIA Seems like this `rope_forward_triton` introduced some accuracy issues?

**Comment by Dawn952** [comment]
**Automated Response**

Please provide: (1) full vLLM startup configuration (--model, --tensor-parallel-size, --max-model-len, --dtype, all flags), (2) complete accuracy comparison data with reproduction steps, (3) NPU count, driver version, vllm-ascend version/commit id, and PLOG logs. @ZYang6263 please follow up.

**Comment by Frozen7771** [comment]
> **Automated Response**
> 
> Please provide: (1) full vLLM startup configuration (--model, --tensor-parallel-size, --max-model-len, --dtype, all flags), (2) complete accuracy comparison data with reproduction steps, (3) NPU count, driver version, vllm-ascend version/commit id, and PLOG logs. [@ZYang6263](https://github.com/ZYang6263) please follow up.

Thanks for following up. Below are the sanitized reproduction details requested.

## 1. Startup configuration

Environment:

- Hardware: Ascend A3, 16 dies
- Driver: 25.5.1
- CANN: 9.0.1
- Python: 3.12.13
- PyTorch: 2.10.0
- torch_npu: 2.10.0.post2
- vLLM: 0.23.0
- vLLM Ascend package identifier: `dev994+gfabc77a96`
- DSpark source revision: `fabc77a96978972c9be0341bd504fabf065abe20`
- triton-ascend: 3.2.1
- Image: `quay.io/ascend/vllm-ascend:nightly-releases-v0.23.0-a3`

The target model is GLM-5.2 W8A8. The draft is our custom DSpark checkpoint using a Qwen3 transformer-layer configuration.

No explicit `--dtype` flag was passed. The draft checkpoint top-level configuration contains:

```text
architectures=["DSparkDraftModel"]
speculators_model_type="dspark"
dtype="bfloat16"
```

Its nested transformer configuration contains five layers, 64 attention heads and `head_dim=192`, with no separate dtype override.

The target model is passed as the positional argument to `vllm serve`; no separate `--model` flag was used.

```bash
export ASCEND_RT_VISIBLE_DEVICES=0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15
export PYTORCH_NPU_ALLOC_CONF=expandable_segments:True
export HCCL_OP_EXPANSION_MODE=AIV
export HCCL_BUFFSIZE=512
export OMP_NUM_THREADS=1
export PYTHONUNBUFFERED=1
export VLLM_ASCEND_ENABLE_GLM_DSPARK_DRAFT_GRAPH=0

SPEC='{"method":"dflash","model":"<DRAFT_MODEL_PATH>","num_speculative_tokens":7,"parallel_drafting":true}'

vllm serve <TARGET_MODEL_PATH> \
  --host 0.0.0.0 \
  --port 18000 \
  --served-model-name glm-52 \
  --trust-remote-code \
  --data-parallel-size 2 \
  --tensor-parallel-size 8 \
  --enable-expert-parallel \
  --enforce-eager \
  --no-enable-prefix-caching \
  --max-model-len 4096 \
  --max-num-seqs 16 \
  --max-num-batched-tokens 8192 \
  --gpu-memory-utilization 0.98 \
  --speculative-config "$SPEC"
```

All diagnostic override and tensor-dump environment variables were unset for the final production comparison.

## 2. Accuracy comparison and reproduction

Reproduction procedure:

1. Start a fresh server using the original dispatched call:

```python
q, k = self.rotary_emb(positions, q, k)
```

2. Run fixed greedy GSM8K requests and measure the speculative-decoding counters strictly before and after:
   - 8 requests at concurrency 1
   - 32 requests at concurrency 8

3. Start another fresh server with the only correctness change:

```python
q, k = self.rotary_emb.forward_native(positions, q, k)
```

4. Repeat the same requests and counter measurements.

Acceptance-length results:

- 8 requests, concurrency 1: `3.055556` before, `5.547692` after
- 32 requests, concurrency 8: `2.910313` before, `5.429936` after

Full32 counter comparison:

- Before: `drafts=7883`, `draft_tokens=55181`, `accepted=15059`, `accept_len=2.910313`
- After: `drafts=4189`, `draft_tokens=29323`, `accepted=18557`, `accept_len=5.429935545476248`

Full32 prefix survival for draft positions 1–7:

```text
Before:
0.718001 / 0.495370 / 0.328682 / 0.188887 / 0.105163 / 0.050108 / 0.024102

After:
0.900692 / 0.805920 / 0.711387 / 0.619480 / 0.535689 / 0.462640 / 0.394127
```

All 40 post-fix requests returned HTTP 200, and the post-run health check remained successful.

A same-worker comparison using the same rotary object, positions, rotary cache and pre-RoPE tensors showed:

- Q dispatched vs native: relative L2 `0.678912`; dispatched post/pre norm ratio `0.736589`
- K dispatched vs native: relative L2 `0.694659`; dispatched post/pre norm ratio `0.732088`

The original inputs and their `.contiguous()` clones produced bitwise-identical dispatched results, excluding input stride or contiguity as the cause. `forward_native()` preserved the Q/K norms.

For the fixed failure at anchor 279 and position 90:

```text
Dispatched result: token 3491
Target top-1:       token 6145
Native result:      token 6145
```

The same result was reproduced in two independent fresh server processes.

## 3. PLOG and runtime logs

This was a silent numerical correctness failure rather than a startup or runtime crash:

- Requests completed successfully.
- The service remained healthy.
- No Python exception or HTTP failure accompanied the incorrect Q/K values.

The relevant container-side PLOG time window and server log were collected. No matching host-side PLOG files were found for the requested time window. The raw container logs contain machine-specific paths and environment information, so they must be sanitized before being attached publicly.

@ZYang6263, I can attach the sanitized PLOG excerpt, server-log excerpt, same-worker comparison JSON and the two complete acceptance-report JSON files for follow-up.

**Comment by ZT-AIA** [comment]
Sorry, I had something to attend to earlier. I'll take a look at it over the next couple of days.

**Comment by ZT-AIA** [comment]
---

## Root Cause & Fix

### Bug

In `vllm_ascend/ops/triton/rope.py`, both `_triton_rope` and `_triton_rope_siso` computed the sin cache offsets using the **padded** rotary dim instead of the real one:

```python
sin_offsets = tl.arange(pad_rope_dim // 2, pad_rope_dim)   # ❌
```

When `rope_dim` is **not a power of two** (e.g. `rope_dim=192`, `pad_rope_dim=next_power_of_2(192)=256`), the sin segment physically starts at `rope_dim/2 = 96`, but the code used `pad_rope_dim/2 = 128` — a **32-element start offset mismatch**, with the tail additionally running past the cache row boundary into garbage. As a result `sin_row` effectively degenerates to ~0/garbage, RoPE collapses to `out = in * cos`, the norm ratio drops to ~√0.5 ≈ 0.707 (issue observed 0.73), attention logits are scaled down, and spec-decode accept_len crashes from 5.6 to 3.0.

### Why `cos_offsets` is fine (same issue, but masked)

`cos_offsets = tl.arange(0, pad_rope_dim // 2)` also uses the padded value, so strictly speaking it has the *same* off-by-pad issue. **But** the cos segment physically starts at `0` regardless of `rope_dim`, so the first `rope_dim/2` elements are always correctly positioned; the excess `[rope_dim/2, pad_rope_dim/2)` lands on sin-segment data but is filtered out by `cos_mask = cos_offsets < (rope_dim // 2)`. Net effect: cos is correct. The sin line, however, has a non-zero start (`rope_dim/2`) that depends on `rope_dim`, so using `pad_rope_dim/2` there shifts the start — mask can't rescue a wrong start.

### Fix (PR #12963, 3 lines)

```diff
-        sin_offsets = tl.arange(pad_rope_dim // 2, pad_rope_dim)
+        sin_offsets = cos_offsets + (rope_dim // 2)
```

Anchor the sin start to the **real** `rope_dim/2`; the existing mask still covers the tail. For power-of-two `rope_dim` (e.g. 128, where `pad_rope_dim == rope_dim`) the new expression is identical to the old one — zero behavior change for common shapes.

### Test

Added `(head_size=128, rotary_dim=96)` (non-power-of-two) to cover this path.

---

Sources:
- [PR #12963 - Fix RoPE cache offsets for non-power-of-two rotary dims](https://github.com/vllm-project/vllm-ascend/pull/12963)
